### PR TITLE
support testing secureboot in Qemu

### PR DESCRIPTION
**Pull request by jepio**
@sayanchowdhury + @jepio

* add new kola command line flags: --qemu-ovmf-vars and --enable-secureboot. --enable-secureboot will eventually be useful for other platforms as well (AWS/GCP/Azure).
* --qemu-bios has been renamed to --qemu-firmware, but --qemu-bios is kept around so that we don't break old releases if we only backport kola changes.
* ovmf vars file is copied into instance folder per test case execution
* skip falco and zfs tests when secure boot: unsigned kernel modules can't be loaded when secure boot is enabled, enforced by lockdown integrity mode.
* force boot from primary disk using `bootindex`

Tested here: http://jenkins.infra.kinvolk.io:8080/job/container/job/test_dispatcher/5120/cldsv/ with ghcr.io/flatcar/mantle:pr-554

### CODE REVIEW

File: kola/harness.go [hunk 1]
```diff
@@ -223,6 +223,13 @@ var (
 	}
 )
 
+func SkipSecureboot(_ semver.Version, channel, arch, platform string) bool {
+	if Options.EnableSecureboot {
+		return true
+	}
+	return false
```
[the remark this answers is not in the document]
Comment (jepio) [reply kola/harness.go:230]: 😂

File: cmd/kola/options.go [hunk 2]
```diff
@@ -222,8 +223,10 @@ func init() {
 	// QEMU-specific options
 	sv(&kola.QEMUOptions.Board, "board", defaultTargetBoard, "target board")
 	sv(&kola.QEMUOptions.DiskImage, "qemu-image", "", "path to CoreOS disk image")
-	sv(&kola.QEMUOptions.BIOSImage, "qemu-bios", "", "BIOS to use for QEMU vm")
+	sv(&kola.QEMUOptions.Firmware, "qemu-bios", "", "bios to use for QEMU vm (deprecated in favor of --qemu-firmware)")
```
[the remark this answers is not in the document]
Comment (jepio) [reply cmd/kola/options.go]: great idea, done

File: kola/harness.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jepio) [reply kola/harness.go]: done

File: kola/tests/sysext/zfs.go [hunk 3]
```diff
@@ -155,10 +156,20 @@ func init() {
...
 }
 
+func skipZfs(version semver.Version, channel, arch, platform string) bool {
+	if kola.SkipSecureboot(version, channel, arch, platform) {
+		return true
+	}
+	if skipOnGha(version, channel, arch, platform) {
+		return true
+	}
+	return false
```
[the remark this answers is not in the document]
Comment (jepio) [reply kola/tests/sysext/zfs.go]: done

File: platform/qemu.go [hunk 4]
```diff
@@ -293,7 +296,28 @@ func mkpath(basedir string) (string, error) {
...
+
+	ovmfVarsCopy, err := os.Create(ovmfVarsDst)
+	if err != nil {
+		return "", err
+	}
+	defer ovmfVarsCopy.Close()
+
+	if _, err := io.Copy(ovmfVarsCopy, ovmfVarsSrc); err != nil {
+		os.Remove(ovmfVarsCopy.Name())
+		return "", err
```
Comment (krnowak) [inline platform/qemu.go:315]: There is a `CopyRegularFile` function in `github.com/flatcar/mantle/system` package. Maybe you could use it instead?

File: platform/qemu.go [hunk 5]
```diff
@@ -349,6 +381,18 @@ func CreateQEMUCommand(board, uuid, biosImage, consolePath, confPath, diskImageP
...
 		"-device", "virtio-rng-pci,rng=rng0",
 	)
+	if ovmfVars != "" {
+		qmCmd = append(qmCmd,
+			"-drive", fmt.Sprintf("if=pflash,unit=0,file=%v,format=raw,readonly=on", firmware),
+			"-drive", fmt.Sprintf("if=pflash,unit=1,file=%v,format=raw", ovmfVars),
+		)
+		if enableSecureboot {
+			qmCmd = append(qmCmd,
+				"-global", "ICH9-LPC.disable_s3=1",
```
Comment (krnowak) [inline platform/qemu.go:391]: This one could use a comment, really.

File: build [hunk 6]
```diff
@@ -30,7 +30,7 @@ cross_build() {
 		echo "Building $a/$1"
 		mkdir -p "bin/$a"
 		CGO_ENABLED=0 GOARCH=$a \
-			go build -mod=vendor -ldflags "${ldflags}" \
+			go build -mod=vendor -ldflags "${ldflags}" -gcflags="all=-N -l" \
```
[the remark this answers is not in the document]
Comment (jepio) [reply build:33]: @sayanchowdhury ?

[the remark this answers is not in the document]
Comment (jepio) [reply build]: removed this again

File: platform/qemu.go (hunk 4 above)
Comment (jepio) [reply platform/qemu.go]: i was looking for something like this. Switched to `CopyRegularFile`

File: platform/qemu.go (hunk 5 above)
Comment (jepio) [reply platform/qemu.go:385]: added a comment with what i know

**Review by krnowak — APPROVED** [review]

**Comment by jepio** [comment]
> Is this still accurate:
> 
> https://github.com/flatcar/mantle/blob/aad4b72a232f6ad03fa26980b5fdc5c9ac68d782/README.md?plain=1#L86
> 
> ? (except the renaming from `qemu-bios` to `qemu-firmware` ?)

Yes, the instructions are still accurate. `--qemu-ovmf-vars` is optional and `--qemu-bios` is still supported. I've pushed a commit updating the README though, so that we have an example of the newest options. I tested both old and new versions.

**Pull request merged**